### ISSUE TITLE
Make Twisted work with latest zope.interface

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2160]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#2160|
|Type|enhancement|
|Created|2006-10-09 17:27:29Z|

Two things changed:

1. There is now a better API for checking whether an adapter has already been registered, and the hack we're using in twisted.python.components no longer works

```
        factory = self.get(origInterface).selfImplied.get(interfaceClass, {}).get('')
```

2. It is no longer possible to override the !__adapt!__ method on interfaces, and web2.iweb.IResource does this. It should instead override !__call!__ or register an adapter hook.

<details><summary>Searchable metadata</summary>

```
trac-id__2160 2160
type__enhancement enhancement
reporter__radix radix
priority__highest highest
milestone__Core_2_5 Core-2.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1160414849000000 1160414849000000
changetime__1160760277000000 1160760277000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Changes in zisvn-2160.  I only tested against Zope3 trunk@HEAD.  Hopefully it still works with actual releases of zope interface, though.  Someone ought to test that.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @exarkun|
|-|-|

Okay, that didn't work.  Test coverage for this area is really crummy. :/

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner|
|-|-|

Okay I think I did it.  Review in zisvn-2160

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @exarkun|
|-|-|

Looks fine, the new tests are much more complete. I've tested with both latest release and trunk and didn't see any problem.

Maybe a comment around the definition of _registered would be welcome, though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [18455]) Merge zisvn-2160

Author: exarkun, radix
Reviewer: therve
Fixes [#2160](https://github.com/twisted/twisted/issues/2160)

Implement the adapter registry using the public AdapterRegistry.registered
method if it is available, otherwise fall-back to the old method which relies
on private Zope Interface APIs.

Also greatly improve test coverage for the adapter registry.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
